## 📍 Agenda 📍
<br>

- [Documentação oficial - Numpy](https://numpy.org/doc/stable/user/whatisnumpy.html)
- [Exercícios](#dois)

### Exercícios

1. Importe numpy como 'np' e imprima o número da versão instalada no ambiente (local ou colab).

2. Crie um *array* 1D com números de 0 a 9 e imprima o resultado.

In [22]:
import numpy as np 
myArray = np.arange(10) 
print(myArray)

[0 1 2 3 4 5 6 7 8 9]


3. Crie uma matriz **booleana** de 3x3 e preencha os seus valores com 'True'. Imprima o resultado.

**Dica:** Verifique o uso da função **full** na documentação do [numpy](https://numpy.org/doc/stable/user/whatisnumpy.html).

In [15]:
import numpy as np
matriz = np.full((3, 3), True)
matriz

array([[ True,  True,  True],
       [ True,  True,  True],
       [ True,  True,  True]])

In [21]:
# dtype tem que ser booleano
e=np.zeros((3,3), dtype=bool)
# 0 == False
# 1 == True
e[e==0] = True
e

array([[ True,  True,  True],
       [ True,  True,  True],
       [ True,  True,  True]])

In [23]:
e=np.zeros((3,3), dtype=bool)
e[e==0] = np.True_
e

array([[ True,  True,  True],
       [ True,  True,  True],
       [ True,  True,  True]])

In [ ]:
np.full((2, 2), np.inf)

In [10]:
u = np.arange(9).reshape(3,3)
troca = np.where(u/1 !=0,  True, True) 
print(troca)

[[ True  True  True]
 [ True  True  True]
 [ True  True  True]]


4. Crie um *array* 1D como números inteiros (50 elementos) de 0 a 1000 aleatórios e **extraia** apenas os números pares. Imprima o resultado.

In [51]:
import numpy as np
arr = np.random.randint(0, 1001, 50)
pares = arr[arr % 2 == 0]
pares

array([122, 362, 288, 862, 176, 270, 460,  30, 370, 726, 912, 688, 650,
       424, 760, 616, 408, 596, 682, 736,   6, 704, 950, 308, 574, 660,
       410, 910, 232, 938, 798, 782, 492])

5. Usando o *array* criando no exercício 4, **substitua** todos os números ímpares maiores que 50 por -1 no *array* original. 

In [52]:
arr[(arr > 50) & (arr % 2 != 0)] = -1
arr

array([ -1, 122,  -1, 362, 288,  -1,  -1, 862, 176,  -1, 270, 460,  -1,
        30, 370, 726,  21, 912, 688, 650,  -1, 424, 760, 616, 408,  -1,
       596, 682, 736,   6,  -1, 704, 950,  -1,  -1, 308, 574, 660, 410,
       910,  -1, 232,  -1, 938, 798, 782,  -1,  -1, 492,   9])

6. Usando o *array* criando no exercício anterior, **substitua** todos os números ímpares com o valor -1 **sem** alterar o *array* original.

In [57]:
import numpy as np
arr = np.random.randint(0, 1001, 50)
arr2 = np.where(arr % 2 == 0, arr, -1)
arr2

array([ -1,  -1,  -1, 524, 838, 626,  -1,  -1, 536, 766,  -1,  -1,  -1,
       622, 136, 310,  -1, 554, 930, 550,  -1,  -1,  70,  -1, 122, 836,
       700, 488,  -1,  -1,  -1, 684, 334, 634, 520,  -1, 692, 948,  -1,
       682,  -1, 526,  -1, 384,  -1, 420, 640, 332, 196, 370])

In [54]:
arr

array([651, 974, 976, 431,  88, 304, 476,  55, 825, 549, 761, 978, 671,
       634, 253, 320,   0, 385, 234, 876, 217,  31, 105, 596,  48, 585,
       645, 847, 668, 878, 126, 861, 462, 369, 893, 440, 552, 662, 473,
       989, 267, 207, 191, 872, 778,  70, 468, 245, 666, 322])

In [50]:
# np.where(arr % 2 == 0, arr, -1)
for i in range(arr.shape[0]):
  if arr[i] % 2 == 0:
    print(arr[i])
  else:
    print(-1)

-1
-1
-1
62
-1
648
954
228
4
-1
-1
836
428
786
-1
-1
830
-1
880
-1
-1
-1
764
-1
434
700
-1
-1
414
806
980
454
492
-1
-1
-1
364
386
516
64
-1
746
-1
-1
148
-1
-1
784
-1
-1


In [46]:
arr

array([858, 704, 443, 347, 550,  59, 446, 701, 319, 408, 904,  97, 511,
       298,  40, 964, 777, 262, 793, 496, 723, 422, 623, 368, 958, 729,
       266, 521, 749, 490,  31, 405, 115, 683, 261,   7, 328, 196, 714,
       446, 711, 798, 322, 497, 490, 876, 180,   9,  93, 727])

In [61]:
import numpy as np
arr = np.random.randint(0, 1001, 50)
arr2 = np.where(arr % 2 == 0, arr, -1)
arr2

array([ 84, 984, 998,  82, 180,  -1,  -1,  -1, 492, 484,  86,  -1, 696,
       960,  -1, 516,  -1,  -1,  -1, 404, 630,  -1, 898, 976,  -1, 338,
       250, 984,  -1,  -1, 294, 664,  -1,  66,  -1, 718, 312,  -1,  38,
       264,  -1,  -1, 802,  -1, 644, 638, 320, 910, 280, 978])

7. Crie um *array* 1D com 100 valores de ponto flutuante (**float**), converta-o em uma matriz 2D com 10 linhas e 10 colunas.

8. Crie duas matrizes 2D de inteiros aleatórios (de 0 a 100) com dimensões 10x10. **Empilhe-as horizontalmente**.

**Dica:** verique as funções **hstack** e **concatenate** na documentação do [numpy](https://numpy.org/doc/stable/user/whatisnumpy.html)

9. Crie duas matrizes 2D de inteiros aleatórios (de 0 a 100) com dimensões 10x10. **Empilhe-as verticalmente**.

Dica: verique as funções **vstack** e **concatenate** na documentação do [numpy](https://numpy.org/doc/stable/user/whatisnumpy.html)


In [35]:
import numpy as np
matriz1 = np.random.randint(0, 101, (10, 10))
matriz2 = np.random.randint(0, 101, (10, 10))

In [36]:
concat_matriz = np.hstack((matriz1, matriz2))

In [39]:
# similar ao hstack
np.concatenate([matriz1, matriz2], axis=1).shape

(10, 20)

In [40]:
# similar ao vstack
np.concatenate([matriz1, matriz2], axis=0).shape

(20, 10)

In [ ]:
# https://stackoverflow.com/questions/33356442/when-should-i-use-hstack-vstack-vs-append-vs-concatenate-vs-column-stack

In [29]:
concat_matriz.shape

(20, 10)

In [37]:
concat_matriz = np.hstack((matriz1, matriz2))

In [31]:
concat_matriz.shape

(10, 20)

Para ler arquivos em formato *txt* que estão em formato de "tabelas", podemos usar a função **genfromtxt()** do **numpy**. Esta função recebe como argumentos: **caminho do arquivo txt** (path local ou uma url); **delimiter** para separar cada valor seguindo algum critério; **dtype** para converter os valores recebidos em algum tipo de dado conhecido (exemplo: **float**) e; **usecols** em que é definido o cabeçalho das colunas lidas a partir do arquivo. 
<br>
<br>
Vejamos um exemplo de uso da função **genfromtxt** sobre a base de dados 
<a href="https://archive.ics.uci.edu/ml/datasets/iris">Iris</a>.

In [ ]:
import numpy as np
# Caminho da base de dados é uma url disponível na internet (visite o link abaixo
# pelo navegador para título de curiosidade :))
url = 'https://archive.ics.uci.edu/ml/machine-learning-databases/iris/iris.data'
# Lê os dados a partir do caminho passado e armazena em uma matriz 2D chamada iris_2d
iris_2d = np.genfromtxt(url, delimiter=',', dtype='float', usecols=[0,1,2,3])

In [ ]:
# matriz 2D criada com os dados da base Iris
iris_2d

array([[5.1, 3.5, 1.4, 0.2],
       [4.9, 3. , 1.4, 0.2],
       [4.7, 3.2, 1.3, 0.2],
       [4.6, 3.1, 1.5, 0.2],
       [5. , 3.6, 1.4, 0.2],
       [5.4, 3.9, 1.7, 0.4],
       [4.6, 3.4, 1.4, 0.3],
       [5. , 3.4, 1.5, 0.2],
       [4.4, 2.9, 1.4, 0.2],
       [4.9, 3.1, 1.5, 0.1],
       [5.4, 3.7, 1.5, 0.2],
       [4.8, 3.4, 1.6, 0.2],
       [4.8, 3. , 1.4, 0.1],
       [4.3, 3. , 1.1, 0.1],
       [5.8, 4. , 1.2, 0.2],
       [5.7, 4.4, 1.5, 0.4],
       [5.4, 3.9, 1.3, 0.4],
       [5.1, 3.5, 1.4, 0.3],
       [5.7, 3.8, 1.7, 0.3],
       [5.1, 3.8, 1.5, 0.3],
       [5.4, 3.4, 1.7, 0.2],
       [5.1, 3.7, 1.5, 0.4],
       [4.6, 3.6, 1. , 0.2],
       [5.1, 3.3, 1.7, 0.5],
       [4.8, 3.4, 1.9, 0.2],
       [5. , 3. , 1.6, 0.2],
       [5. , 3.4, 1.6, 0.4],
       [5.2, 3.5, 1.5, 0.2],
       [5.2, 3.4, 1.4, 0.2],
       [4.7, 3.2, 1.6, 0.2],
       [4.8, 3.1, 1.6, 0.2],
       [5.4, 3.4, 1.5, 0.4],
       [5.2, 4.1, 1.5, 0.1],
       [5.5, 4.2, 1.4, 0.2],
       [4.9, 3

11. A partir dos dados lidos na matriz iris_2d, filtre as linhas que possuem (coluna 3) > 1.5 e (coluna 1) < 5.0.

12. A partir dos dados lidos na matriz iris_2d, calcule a **média** e **desvio padrão** de cada coluna presente nos dados. Imprima o resultado.

13. Crie um repositório no Github e versione todos os códigos que desenvolveram nesta lista de exercícios.